### PR TITLE
fix(udf): prevent continuous retries on connection error

### LINKED ISSUE #13791: UDF retry should record the error state
We add the UDF retry mechanism in #13375 , however it's dangerous.

In simple terms, we introduced the non-strict evaluation to prevent a single request from blocking the entire stream. When encountering UDF errors, we will fill in NULL values.

The current retry mechanism is chunk-based, which means that when the UDF server is down and can’t be recovered soon, every chunk will introduce about 3s block, and the back pressure will increase uncontrollably! This goes against the original intention of non-strict evaluation.

The fix should be cherry-picked.

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #13791 by implementing https://github.com/risingwavelabs/risingwave/issues/13791#issuecomment-1839261633

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

### CODE REVIEW

File: src/udf/src/error.rs [hunk 1]
```diff
@@ -52,12 +52,6 @@ impl Error {
     /// Returns true if the error is caused by a connection error.
     pub fn is_connection_error(&self) -> bool {
         match self.inner() {
-            // stream closed because of a broken pipe
-            ErrorInner::Flight(FlightError::Tonic(status))
-                if status.code() == tonic::Code::Unknown =>
-            {
-                true
-            }
```
Comment (wangrunji0408) [inline src/udf/src/error.rs:60]: Python/Java exceptions in user functions will fall into this branch. So remove it from connection error.

**Comment by TennyZhuang** [comment]
I am pondering whether we can consider one success as representing not being in a failing state, which seems unable to handle crashed servers hidden behind load balancing. I suggest using a magic number like 16.

File: src/expr/core/src/expr/expr_udf.rs [hunk 2]
```diff
@@ -39,12 +39,16 @@ pub struct UdfExpression {
...
     span: await_tree::Span,
-    /// Whether a connection error has occurred in the last call.
-    /// If true, we will not retry subsequent calls to prevent blocking the stream.
+    /// Number of remaining successful calls until retry is enabled.
+    /// If non-zero, we will not retry on connection errors to prevent blocking the stream.
+    /// On each connection error, the count will be reset to `INITIAL_RETRY_COUNT`.
+    /// On each successful call, the count will be decreased by 1.
     /// See <https://github.com/risingwavelabs/risingwave/issues/13791>.
-    failing: AtomicBool,
+    disable_retry_count: AtomicU8,
```
Comment (wangrunji0408) [inline src/expr/core/src/expr/expr_udf.rs:47]: I can't think of a proper name for this variable. Any suggestions are welcome.

Comment (BugenZhao) [reply src/expr/core/src/expr/expr_udf.rs:47]: What about using the opposite semantics? For example, `continuous_success`.

**Review by BugenZhao — APPROVED** [review]

File: src/expr/core/src/expr/expr_udf.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/core/src/expr/expr_udf.rs:47]: Considering retry enabled is a normal state and retry disabled is a special state, I prefer using the default value 0 to represent the normal state. 🤔

**Review by TennyZhuang — APPROVED** [review]

**Pull request merged**